include/revng/PTML/CommentPlacementHelper.h

```diff
@@ -0,0 +1,354 @@
+///   //       `decltype(MyGraph) *` OR provide a specialization of
+///   //       `llvm::DomTreeNodeTraits`.
+///
+///   // 2. When emitting your artifact, add calls to `getComments`:
+///   for (const auto &Statement : myWayOfIteratingTheGraph(MyGraph)) {
+///     // . . .
+///     for (const auto &Comment : CommentPlacement.getComments(Statement)) {
+///       emitTheComment(Comment);
+///       // note: see `PTML/Doxygen` for the standard way of emitting comments
+///     }
+///     // . . .
+///   }
+///
+///   // 3. Don't forget to emit homeless comments somewhere.
+///   for (const auto &Comment : CommentPlacement.getHomelessComments())
+///     emitTheComment(Comment);
+/// ```
+///
+/// With the interface out of the way, lets now discuss _how_ statements are
+/// selected:
+///
+/// 1. For each comment-statement pair a similarity score is computed
+///    (see \ref tverskyIndex) by traversing the graph.
+/// 2. Then, for each comment a single _address set_ (note: not a statement yet)
+///    is selected.
+/// 3. Then graph is traversed again, this time looking for all the statements
+///    whose location exactly matches the selected one, except for statements
+///    that are reachable in a dominator tree from an already selected
+///    statement.
+/// 4. All these statements are recorded in a \ref ResultMap and will be emitted
+///    once requested by a \ref getComments call.
+template<HasStatementTraits NodeType>
+class CommentPlacementHelper {
+private:
+  using Trait = StatementTraits<NodeType>;
+  using InternalNodeType = std::remove_pointer_t<NodeType>;
+
+  using StatementComment = model::StatementComment;
+  using CLT = decltype(std::declval<StatementComment>().Location());
+  using CommentLocationType = std::decay_t<CLT>;
+
+  using StatementType = typename Trait::StatementType;
+  using SLT = decltype(Trait::getAddresses(std::declval<StatementType>()));
+  using StatementLocationType = std::decay_t<SLT>;
+
+  struct Score {
+    uint64_t Numerator = 0;
+    uint64_t Denominator = 1;
+
+    std::strong_ordering operator<=>(const Score &Another) const {
+      // This takes advantage of the easiest way to compare two fractions:
+      // instead of looking for the _smallest_ common denominator,
+      // just use _one_ common denominator: product of the two denominators.
+      //
+      // Say, we're comparing `a/b` and `c/d`: _one_ common denominator is
+      // `b*d`, which means we can just compare `(a*d)/(b*d)` and `(b*c)/(b*d)`.
+      // And since "adjusted" denominators are the same, we can also omit them
+      // and just compare the numerators.
+      uint64_t AdjustedLHS = this->Numerator * Another.Denominator;
+      uint64_t AdjustedRHS = Another.Numerator * this->Denominator;
+      return AdjustedLHS <=> AdjustedRHS;
+    }
+    bool operator==(const Score &Another) const {
+      return operator<=>(Another) == std::strong_ordering::equal;
+    }
+  };
+
+public:
+  struct CommentView {
```

Comment:
In what sense is this a *view*? Maybe `CommentPlacement` or `CommentAssignment`?

Reply:
It the same sense `string_view` is a view. As in a smart "reference" to something owned elsewhere. That being said, I have no qualms about renaming it.

Reply:
Done (`CommentAssignment`).